File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/text/MessageArguments.java
package top.likoslupus.cellulosesz.api.text;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import static java.util.Objects.requireNonNull;

/** Immutable ordered collection of typed message arguments. */
public final class MessageArguments {

    private static final MessageArguments EMPTY = new MessageArguments(List.of());

    private final List<MessageArgument> values;

    private MessageArguments(List<MessageArgument> values) {
        this.values = List.copyOf(requireNonNull(values, "values"));
    }

    public static MessageArguments empty() {
        return EMPTY;
    }

    public static MessageArguments of(MessageArgument... arguments) {
        requireNonNull(arguments, "arguments");
        if (arguments.length == 0) {
            return EMPTY;
        }
        return new MessageArguments(List.of(arguments));
    }

    public static Builder builder() {
        return new Builder();
    }

    public List<MessageArgument> values() {
        return values;
    }

    public boolean isEmpty() {
        return values.isEmpty();
    }

    public static final class Builder {

        private final List<MessageArgument> values = new ArrayList<>();

        private Builder() {
        }

        public Builder add(String value) {
            return add(MessageArgument.text(value));
        }

        public Builder add(MessageArgument value) {
            values.add(requireNonNull(value, "value"));
            return this;
        }

        public Builder add(int value) {
            return add(MessageArgument.number(value));
        }

        public Builder add(long value) {
            return add(MessageArgument.number(value));
        }

        public Builder add(double value) {
            return add(MessageArgument.number(value));
        }

        public Builder add(boolean value) {
            return add(MessageArgument.bool(value));
        }

        public Builder add(BigDecimal value) {
            return add(MessageArgument.number(value));
        }

        public Builder add(UUID value) {
            return add(MessageArgument.uuid(value));
        }

        public Builder add(RichText value) {
            return add(MessageArgument.richText(value));
        }

        public Builder add(LocalizedMessage value) {
            return add(MessageArgument.nested(value));
        }

        public Builder addAll(MessageArguments arguments) {
            values.addAll(requireNonNull(arguments, "arguments").values);
            return this;
        }

        public MessageArguments build() {
            return values.isEmpty()
                    ? EMPTY
                    : new MessageArguments(values);
        }

    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/text/MessageArguments.kt
package top.likoslupus.cellulosesz.api.text

import top.likoslupus.cellulosesz.api.util.toImmutableList
import java.math.BigDecimal
import java.util.*

/** Immutable ordered collection of typed message arguments. */
public class MessageArguments private constructor(
    private val values: List<MessageArgument>
) {

    public fun values(): List<MessageArgument> = values

    public fun isEmpty(): Boolean = values.isEmpty()

    public class Builder internal constructor() {

        private val values: MutableList<MessageArgument> = ArrayList()

        public fun add(value: String): Builder =
            add(MessageArgument.text(value))

        public fun add(value: MessageArgument): Builder {
            values.add(value)
            return this
        }

        public fun add(value: Int): Builder =
            add(MessageArgument.number(value))

        public fun add(value: Long): Builder =
            add(MessageArgument.number(value))

        public fun add(value: Double): Builder =
            add(MessageArgument.number(value))

        public fun add(value: Boolean): Builder =
            add(MessageArgument.bool(value))

        public fun add(value: BigDecimal): Builder =
            add(MessageArgument.number(value))

        public fun add(value: UUID): Builder =
            add(MessageArgument.uuid(value))

        public fun add(value: RichText): Builder =
            add(MessageArgument.richText(value))

        public fun add(value: LocalizedMessage): Builder =
            add(MessageArgument.nested(value))

        public fun addAll(arguments: MessageArguments): Builder {
            values.addAll(arguments.values)
            return this
        }

        public fun build(): MessageArguments =
            if (values.isEmpty()) {
                EMPTY
            } else {
                MessageArguments(values.toImmutableList())
            }

    }

    public companion object {

        private val EMPTY = MessageArguments(emptyList())

        @JvmStatic
        public fun empty(): MessageArguments = EMPTY

        @JvmStatic
        public fun of(vararg arguments: MessageArgument): MessageArguments =
            if (arguments.isEmpty()) {
                EMPTY
            } else {
                MessageArguments(arguments.toList())
            }

        @JvmStatic
        public fun builder(): Builder = Builder()

    }

}
